Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,222 @@
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security
+{
+    public class NuGetCredentialsValidator : ValidatorBase
+    {
+        internal static NuGetCredentialsValidator Instance;
+
+        // We can't count on the int values of the enumerations being ordered as we want,
+        // so order them manually
+        private readonly Dictionary<ValidationState, int> badResponseSorting = new Dictionary<ValidationState, int>()
+        {
+            { ValidationState.Unknown, 0 },
+            { ValidationState.UnknownHost, 1 },
+            { ValidationState.Unauthorized, 2 },
+        };
+
+        static NuGetCredentialsValidator()
+        {
+            Instance = new NuGetCredentialsValidator();
+        }
+
+        public static IEnumerable<ValidationResult> IsValidStatic(ref string matchedPattern,
+                                                                  ref Dictionary<string, string> groups,
+                                                                  ref string message)
+        {
+            return IsValidStatic(Instance,
+                                 ref matchedPattern,
+                                 ref groups,
+                                 ref message);
+        }
+
+        public static ValidationState IsValidDynamic(ref Fingerprint fingerprint,
+                                                     ref string message,
+                                                     ref Dictionary<string, string> options,
+                                                     ref ResultLevelKind resultLevelKind)
+        {
+            return IsValidDynamic(Instance,
+                                  ref fingerprint,
+                                  ref message,
+                                  ref options,
+                                  ref resultLevelKind);
+        }
+
+        internal static List<string> ExtractHosts(string hostXmlAsString)
+        {
+            try
+            {
+                return ExtractHostsHelper(hostXmlAsString);
+            }
+            catch (XmlException)
+            {
+                // Maybe it's escaped? Try to unescape it...
+                try
+                {
+                    string attemptTwoString = Regex.Unescape(hostXmlAsString).Replace(Environment.NewLine, string.Empty);
+
+                    // Environment.NewLine is \r\n, and it will miss \n alone, so...
+                    attemptTwoString = attemptTwoString.Replace("\n", string.Empty);
+                    return ExtractHostsHelper(attemptTwoString);
+                }
+                catch
+                {
+                    // Still gotta return an empty list even if parsing failed
+                    return new List<string>();
+                }
+            }
+        }
+
+        protected override IEnumerable<ValidationResult> IsValidStaticHelper(ref string matchedPattern,
```

Comment:
>IsValidStaticHelper [](http://example.com/codeflow?start=57&length=19)

This rule needs improvement! There are an arbitrary number of hosts in a config file, but you can also specify an arbitrary number of user credentials!

and so this helper should return host * secrets fingerprint candidates.

[https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file](https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file)

<packageSourceCredentials>
    <Contoso>
        <add key="Username" value="user@contoso.com" />
        <add key="Password" value="..." />
        <add key="ValidAuthenticationTypes" value="basic" />
    </Contoso>
    <Test_x0020_Source>
        <add key="Username" value="user" />
        <add key="ClearTextPassword" value="hal+9ooo_da!sY" />
        <add key="ValidAuthenticationTypes" value="basic, negotiate" />
    </Test_x0020_Source>
</packageSourceCredentials>

Reply:
added as an issue to unblock the merge.